lib/types/record.nix

```diff
@@ -0,0 +1,124 @@
+{ lib }:
+
+let
+  inherit (lib)
+    mapAttrs
+    zipAttrsWith
+    removeAttrs
+    attrNames
+    showOption
+    optional
+    mergeDefinitions
+    mkOptionType
+    isAttrs
+    ;
+
+  record =
+    {
+      fields ? { },
+      optionalFields ? { },
+      wildcard ? null,
+    }@args:
+    # TODO: assert that at least one arg is provided, i.e. args != { }
+    let
+      checkFields = mapAttrs (
+        name: value:
+        if value._type or null != "option" then
+          throw "Record field `${lib.escapeNixIdentifier name}` must be declared with `mkOption`."
+        else if value ? apply then
+          throw "In field ${lib.escapeNixIdentifier name} records do not support options with `apply`"
+        else if value ? readOnly then
+          throw "In field ${lib.escapeNixIdentifier name} records do not support options with `readOnly`"
+        else
+          value
+      );
+      fields' = checkFields fields;
+      # TODO: should we also check optionalFields have no default?
```

Comment:
I agree a `mkField` would make sense to further de-couple records from modules. This would also allow fields to declare whether they are optional, rather than having a separate `optionalFileds` argument.